Firestore/core/src/firebase/firestore/local/leveldb_remote_document_cache.cc

```diff
@@ -102,13 +146,22 @@ OptionalMaybeDocumentMap LevelDbRemoteDocumentCache::GetAll(
     it->Seek(LevelDbRemoteDocumentKey::Key(key));
     if (!it->Valid() || !current_key.Decode(it->key()) ||
         current_key.document_key() != key) {
-      results = results.insert(key, absl::nullopt);
+      results.Insert(std::make_pair(key, absl::nullopt));
```

Comment:
More of a question regarding the consumers of this method: Could we change the callers so that we don't have to insert `null` values (they could instead look at the existence of the keys). This would simplify this method a bit: We would not need to insert `std::pair`s anymore and `AsyncResults` wouldn't need to use templates anymore (or if it did, there would only be one implementation of it).

Reply:
It's possible, but messy. It may not be a net win.

The issue is that we use OptionalMaybeDocumentMaps to convey two pieces of information:

  * That a key was requested/affected
  * That a value exists for a key (and may be a NoDocument)

So for example, the local documents view can take this result and pipe it directly into GetLocalViewOfDocuments, which uses the keys of the map to go fetch affected mutations. Without this, we'd have to pass the keys separately to ensure that we find mutations that apply to documents that weren't in the remote document cache.

Similarly, rejecting a mutation batch uses the return value of GetDocuments to indicate the affected documents in a view. If the result doesn't contain the nullopts for things that were deleted, LocalStore::RejectMutationBatch would have to add them back in.

Adjusting this really becomes an exercise in getting https://github.com/firebase/firebase-js-sdk/pull/1886 right, which seems like more work than is warranted right now.